File: code/ex129.py
import copy
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def helper(self, root, p, mark):
        global path, s
        
        if root != None:
            p = root
            
            val = str(root.val)
            path.append(val)
            
            self.helper(root.left, p, 1)
            self.helper(root.right, p, 2)
            
            #Return to the last node
            path.pop(-1)
        else:
            if (p.left == None) and (p.right == None):
                if mark == 1:
                    s += int(''.join(copy.deepcopy(path)))
                   
        
        
    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        
        global path, s
        path = []
        s = 0
        
        if root == None:
            return s
        
        self.helper(root, root, 0)
        
        return s

File: code/ex129.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int sum = 0;
public:
    int sumNumbers(TreeNode* root) {
        if (root == NULL)    return 0;
        dfs(root, 0);
        return sum;
    }
    
private:
    void dfs(TreeNode* root, int num) {
        num = num * 10 + root->val;
        if (root->left == NULL && root->right == NULL) {
            sum += num;
            return;
        }
        if (root->left != NULL)    dfs(root->left, num);
        if (root->right != NULL)    dfs(root->right, num);
    }
};